subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -2443,4 +2438,42 @@ public Optional<? extends RuntimeException> mapFailure(String type, Collection<T
             return DefaultConfiguration.this.mapFailure(type, failures);
         }
     }
+
+    /**
+     * Helper class that allows all configuration usage information to be stored in a byte using bitwise operations.
+     * <p>
+     * This may slightly trim the memory usage of configuration instances, which can be numerous in larger builds.
+     */
+    private enum Usage {
+        CONSUMABLE((byte) 1),
+        RESOLVABLE((byte) 2),
+        DECLARABLE_AGAINST((byte) 4),
+        DEPRECATED_FOR_CONSUMPTION((byte) 8),
+        DEPRECATED_FOR_RESOLUTION((byte) 16),
+        DEPRECATED_FOR_DECLARATION_AGAINST((byte) 32);
+
+        private static final byte EVERYTHING_ALLOWED_NOTHING_DEPRECATED = 7; // consumable & resolvable & declarable_against
+
+        private final byte mask;
+
+        Usage(byte mask) {
+            this.mask = mask;
+        }
+
+        public byte toggle(boolean allowed, byte currentUsage) {
+            return allowed ? addTo(currentUsage) : removeFrom(currentUsage);
+        }
+
+        private byte addTo(byte currentUsage) {
+            return (byte)(mask | currentUsage);
+        }
+
+        private byte removeFrom(byte currentUsage) {
+            return (byte)(~mask & currentUsage);
+        }
+
+        public boolean isEnabled(DefaultConfiguration configuration) {
+            return (mask & configuration.allowedUsage) > 0;
```

Comment:
```suggestion
            return (mask & configuration.allowedUsage) != 0;
```
If we ever set the sign bit this must be `!=`, better to make sure it doesn't cause surprises in the future.